lib/spack/spack/test/cmd/env.py

```diff
@@ -1921,6 +1921,140 @@ def test_env_include_concrete_remove_env():
     assert test2.path not in lockfile_as_dict["include_concrete"].keys()
 
 
+def configure_reuse_true(env):
+    with env:
+        config("add", "concretizer:reuse:true")
+
+
+def configure_reuse_from_environment(env):
+    env.manifest.configuration.update(
+        {"concretizer": {"reuse": {"from": [{"type": "environment"}]}}}
+    )
+    env.manifest.changed = True
+    env.write()
+
+
+def configure_reuse_from_environment_test1(env):
+    env.manifest.configuration.update(
+        {"concretizer": {"reuse": {"from": [{"type": {"environment": "test1"}}]}}}
+    )
+    env.manifest.changed = True
+    env.write()
+
+
+def configure_reuse_from_environment_external_test(combined_env):
+    # Create a new environment called external_test
+    env("create", "external_test")
+    ext = ev.read("external_test")
+    with ext:
+        add("mpich@1.0 +debug")
+    ext.concretize()
+    ext.write()
+
+    combined_env.manifest.configuration.update(
+        {
+            "concretizer": {"reuse": {"from": [{"type": {"environment": "external_test"}}]}},
+            "packages": {"mpich": {"require": ["+debug"]}},
+        }
+    )
+    combined_env.manifest.changed = True
+    combined_env.write()
+
+    return ext
+
+
+def configure_reuse_from_environment_raise(env):
+    env.manifest.configuration.update(
+        {"concretizer": {"reuse": {"from": [{"type": {"environment": "not-a-real-env"}}]}}}
+    )
+    env.manifest.changed = True
+    env.write()
+
+
+@pytest.mark.parametrize(
+    "configure_reuse",
+    [
+        "true",
+        "from_environment",
+        "from_environment_test1",
+        "from_environment_external_test",
+        "from_environment_raise",
+    ],
+)
+def test_env_include_concrete_reuse(monkeypatch, configure_reuse):
+
+    # The mock packages do not use the gcc-runtime
+    def mock_has_runtime_dependencies(*args, **kwargs):
+        return True
+
+    monkeypatch.setattr(
+        spack.solver.asp, "_has_runtime_dependencies", mock_has_runtime_dependencies
+    )
+    # test1 contains mpich@1, make sure it is reused in the combined environment
+    # Do not reuse specs from external environments
+    # (included concrete environments have different behavior)
+    test1, _, combined = setup_combined_multiple_env()
+
+    # The default mpi version is 3.x provided by mpich in the mock repo.
+    # This test verifies that concretizing with an included concrete
+    # environment with "concretizer:reuse:true" the included
+    # concrete spec overrides the default with mpi@1.0.
+    print(test1.name)
+    print(test1.concrete_roots())
```

Comment:
We shouldn't have any print statements in the test once this is ready to merge -- is there something we need to `assert` here or is this just leftover from debugging?

Reply:
leftover debug, I removed the stuff in `asp.py` but missed it here.